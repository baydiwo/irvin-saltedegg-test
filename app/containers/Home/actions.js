/*
 *
 * Home actions
 *
 */

import {
  DEFAULT_ACTION,
  FETCH_DETAIL,
  FETCH_DETAIL_SUCCESS,
  FETCH_DETAIL_ERROR,
  ADD_TO_CART,
  REMOVE_ITEM,
  ADD_QUANTITY,
  TOTAL_PRICE
} from "./constants";

// flow type
type ActionWithPayloadType<T, P> = {|
  type: T,
  payload: P,
|};

type ActionWithPayloadIdType<T, V, I> = {|
  type: T,
  value: V,
  id: I
|};

type ActionWithoutPayloadType<T> = {|
  type: T,
|};
export type SuccessPayloadType = {|
  products: any,
|};

export type ErrorPayloadType = {|
  error: string,
|};

// action type
type GetProductsSuccessActionType = ActionWithPayloadType<
  typeof FETCH_DETAIL_SUCCESS,
  SuccessPayloadType
>;

type GetProductsErrorActionType = ActionWithPayloadType<
  typeof FETCH_DETAIL_ERROR,
  ErrorPayloadType
>;

type AddToCartActionType = ActionWithPayloadType<
  typeof ADD_TO_CART,
  any
>;

type RemoveItemActionType = ActionWithPayloadType<
  typeof REMOVE_ITEM,
  any
>;

type TotalPriceActionType = ActionWithoutPayloadType<
  typeof TOTAL_PRICE
>;

type AddQuantityActionType = ActionWithPayloadIdType<
typeof ADD_QUANTITY,
number,
string
>;

// actions
export function defaultAction() {
  return {
    type: DEFAULT_ACTION
  };
}

export function fetchDetail() {
  return {
    type: FETCH_DETAIL,
  };
}

export function fetchDetailSuccess(
    products: SuccessPayloadType
  ): GetProductsSuccessActionType {
    return {
      type: FETCH_DETAIL_SUCCESS,
      products,
    };
  }

export function fetchDetailError(error: ErrorPayloadType): GetProductsErrorActionType{
  return {
    type: FETCH_DETAIL_ERROR,
    error,
  };
}

export function addToCart(
         id: any
       ): AddToCartActionType {
         return {
           type: ADD_TO_CART,
           id,
         };
       };
//remove item action
export function removeItem(id: any): RemoveItemActionType {
  return {
    type: REMOVE_ITEM,
    id,
  };
};

//add qt action
export function addQuantity(value: number, id: string) {
  return {
    type: ADD_QUANTITY,
    value,
    id,
  };
};

export function totalPrice(): TotalPriceActionType {
  return {
    type: TOTAL_PRICE,
  };
};