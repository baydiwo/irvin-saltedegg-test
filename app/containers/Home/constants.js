/*
 *
 * Home constants
 *
 */

export const DEFAULT_ACTION = "app/Home/DEFAULT_ACTION";
export const FETCH_DETAIL = "app/Home/FETCH_DETAIL";
export const FETCH_DETAIL_SUCCESS = "app/Home/FETCH_DETAIL_SUCCESS";
export const FETCH_DETAIL_ERROR = "app/Home/FETCH_DETAIL_ERROR";
export const ADD_TO_CART = "app/Home/ADD_TO_CART";
export const REMOVE_ITEM = "app/Home/REMOVE_ITEM";
export const SUB_QUANTITY = "app/Home/SUB_QUANTITY";
export const ADD_QUANTITY = "app/Home/ADD_QUANTITY";
export const TOTAL_PRICE = "app/Home/TOTAL_PRICE";