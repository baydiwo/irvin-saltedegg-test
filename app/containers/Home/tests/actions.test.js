import * as actions from "../actions";
import { DEFAULT_ACTION } from "../constants";

describe("Home actions", () => {
  describe("Default Action", () => {
    it("has a type of DEFAULT_ACTION", () => {
      const expected = {
        type: DEFAULT_ACTION
      };
      expect(actions.defaultAction()).toEqual(expected);
    });

    it("fetchDetailSuccess", () => {
      const payload = {
        products: [
          {
            currency: "SGD",
            image:
              "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
            name: "IPhone 7",
            price: "15",
          },
        ],
      };
      expect(actions.fetchDetailSuccess(payload)).toMatchSnapshot();
    });

    it("fetchDetailError", () => {
      const payload = { error: "Some error message" };
      expect(actions.fetchDetailError(payload)).toMatchSnapshot();
    });

    it("addToCart", () => {
      const payload = {
        products: [
          {
            currency: "SGD",
            image:
              "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
            name: "IPhone 7",
            price: "15",
          },
        ],
      };
      expect(actions.addToCart(payload)).toMatchSnapshot();
    });
    it("addToCart", () => {
      const payload = {
        products: [
          {
            currency: "SGD",
            image:
              "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
            name: "IPhone 7",
            price: "15",
          },
        ],
      };
      expect(actions.addToCart(payload)).toMatchSnapshot();
    });
    
    it("addQuantity", () => {
      
      const value = {
        products: [
          {
            currency: "SGD",
            image:
              "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
            name: "IPhone 7",
            price: "15",
          },
        ],
      };
      expect(actions.addQuantity(value)).toMatchSnapshot();
    });
    
    it("totalPrice", () => {
      expect(actions.totalPrice(200)).toMatchSnapshot();
    });

  });
});
