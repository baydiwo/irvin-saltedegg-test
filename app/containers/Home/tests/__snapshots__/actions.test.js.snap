// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home actions Default Action addQuantity 1`] = `
Object {
  "id": undefined,
  "type": "app/Home/ADD_QUANTITY",
  "value": Object {
    "products": Array [
      Object {
        "currency": "SGD",
        "image": "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
        "name": "IPhone 7",
        "price": "15",
      },
    ],
  },
}
`;

exports[`Home actions Default Action addToCart 1`] = `
Object {
  "id": Object {
    "products": Array [
      Object {
        "currency": "SGD",
        "image": "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
        "name": "IPhone 7",
        "price": "15",
      },
    ],
  },
  "type": "app/Home/ADD_TO_CART",
}
`;

exports[`Home actions Default Action addToCart 2`] = `
Object {
  "id": Object {
    "products": Array [
      Object {
        "currency": "SGD",
        "image": "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
        "name": "IPhone 7",
        "price": "15",
      },
    ],
  },
  "type": "app/Home/ADD_TO_CART",
}
`;

exports[`Home actions Default Action fetchDetailError 1`] = `
Object {
  "error": Object {
    "error": "Some error message",
  },
  "type": "app/Home/FETCH_DETAIL_ERROR",
}
`;

exports[`Home actions Default Action fetchDetailSuccess 1`] = `
Object {
  "products": Object {
    "products": Array [
      Object {
        "currency": "SGD",
        "image": "http://s3-ap-southeast-1.amazonaws.com/s3.irvinsaltedegg.com/engineering-test/images/product-1.jpg",
        "name": "IPhone 7",
        "price": "15",
      },
    ],
  },
  "type": "app/Home/FETCH_DETAIL_SUCCESS",
}
`;

exports[`Home actions Default Action totalPrice 1`] = `
Object {
  "type": "app/Home/TOTAL_PRICE",
}
`;
